src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,551 @@
+        impressions, of which k are shown.
+
+        Args:
+          k:  np.array specifying number of impressions that were seen by the user.
+          p:  Probability that a given impression will be chosen.
+        Returns:
+          For each k, probability that a randomly chosen user will have an inventory
+          of n impressions, of which k are shown.
+        """
+        return np.array(
+            [np.sum([self.knreach(kv, np.arange(kv, len(self._pmf)), p)]) for kv in k]
+        )
+
+    def kplusreach(self, k, p):
+        """Probability that a random user receives k or more impressions.
+
+        Computed as one minus the probability that the user receives k-1 or
+        fewer impressions.
+
+        Args:
+          k:  np.array specifying number of impressions that were seen by the user.
+          p:  Probability that a given impression will be chosen.
+        Returns:
+          For each k, probability that a randomly chosen user will have an inventory
+          of n impressions, of which k are shown.
+        """
+        if k == 0:
+            return 1.0
+        else:
+            return 1.0 - np.sum(self.kreach(np.arange(k), p))
+
+    def expected_value(self):
+        """Expected value of this distribution."""
+        return np.sum(np.arange(len(self._pmf)) * self._pmf)
+
+
+class KInflatedGammaPoissonModel(ReachCurve):
+    """k-inflated Gamma-Poisson reach curve model for underreported counts.
+
+    The k-inflated Gamma-Poisson distribution is like the Gamma-Poisson
+    distribution except that the first k values of the PMF are allowed to
+    be arbitrary.
+
+    It is given by k+3 parameters: alpha, beta, N, a_1, a_2, ..., a_k.
+    We assume that each a_i >= 0, and that a_1 + a_2 + ... + a_k < 1.
+
+    Let GP(n | alpha, beta) be the PMF of the shifted Gamma-Poisson distribution
+    with parameters alpha, beta.  Also, define s = a_1 + a_2 + ... + a_k,
+    and define t = \sum_{i=1}^k GP(i | alpha, beta).  Then, the PMF of the
+    k-inflated Gamma-Poisson distribution is given as
+
+     kGP(n | alpha, beta, a_1, ..., a_k)
+          = a_n if n <= k,
+            (1-s) / (1-t) GP(n | alpha, beta) if n > k.
+
+    As with the Gamma-Poisson distribution, we use the chi2 objective
+    function for assessing goodness of fit.  That is to say, let h[i]
+    be the number of people that were reached i times, and let hbar[i]
+    be the estimated number of people that would be reached i times,
+    given alpha, beta and N.  The objective function that we
+    compute is as follows:
+
+      chi2(h, hbar) = \sum_i (h[i] - hbar[i])^2 / hbar[i]
+
+    One question is what value of k should be chosen.  We propose
+    to consider successive values of k starting from 0, and to stop
+    when a fit is found such that the value of the objective function
+    is smaller than the value of the chi^2 statistic with f_max degrees
+    of freedom, where f_max is the highest frequency that is recorded.
```

Comment:
Then, can k ever stop at or near fmax (or precisely, the min fmax across training points)?  In that case, the model would seem too flexible and have little to do with Gamma Poisson.

Reply:
Noticed that you set a kmax below.  I'd like to print out (k_opt, fmax, kmax) for comparison.  If you've investigated into it, would you please show me some examples?

Reply:
I have done a number of ad hoc tests, but I don't have a table of values that I can share with you, although I could put one together if you like.  You are right that k can become arbitrarily large.  In fact, when I have tried it with data that is clearly not Gamma-Poisson, I have often observed that the algorithm terminates when k = f_max.  I would not say that the model is "too flexible".  Rather, I would say that the model becomes non-parametric.  In other words, when the data is close to a Gamma-Poisson distribution, then the algorithm will terminate with a low value of k.  But if the data is not close to a Gamma-Poisson, then the model acts like a non-parametric model.  So, in that sense, this model interpolates between these two types of models -- the parameter Gamma-Poisson vs. a fully non-parametric model.  I look forward to hearing your thoughts.